### PR TITLE
[WIP] Implemented model observers

### LINKED ISSUE #6: Integrate model events
http://laravel.com/docs/5.0/eloquent#model-events

### PR DESCRIPTION
**Pull request by wouterj**
/fixes #6, #42 

Todo
---

 * [x] Allow observers to be services
 * [x] Docs

### CODE REVIEW

File: tests/Functional/EventsTest.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                $table->string('password');
+                $table->timestamps();
+            });
+        }
+    }
+
+    /** @dataProvider getEventsData */
+    public function testEvents($setup, $execute, array $expectedEvents)
+    {
+        ob_start();
```
Comment (theofidry) [inline tests/Functional/EventsTest.php]: is it necessary?

Comment (wouterj) [inline tests/Functional/EventsTest.php]: Nope, just back from a holiday and my brain isn't thinking 100% straight yet ;) Used a better technique now.

**Comment by wouterj** [comment]
Ready to merge imo. Will wait for a day or 2 before merging and releasing 0.4.0. If anyone has some time to review, please do.

File: src/Events/ServiceContainerDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Illuminate\Events\Dispatcher;
+
+class ServiceContainerDispatcher extends Dispatcher
+{
+    private $symfonyContainer;
+    private $observers;
+
+    public function __construct(ContainerInterface $symfonyContainer, array $observers)
```
Comment (theofidry) [inline src/Events/ServiceContainerDispatcher.php]: missing typehint

File: tests/Functional/Events_ClassTest.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+namespace WouterJ\EloquentBundle\Functional;
+
+use AppBundle\Model\UserObserver;
+
+/**
+ * @author Wouter J <wouter@wouterj.nl>
+ */
+class Events_ClassTest extends EventsTest
```
Comment (theofidry) [inline tests/Functional/Events_ClassTest.php]: why the partial snake case here?

File: tests/Functional/Events_ServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+namespace WouterJ\EloquentBundle\Functional;
+
+use AppBundle\Model\UserObserver;
+
+/**
+ * @author Wouter J <wouter@wouterj.nl>
+ */
+class Events_ServiceTest extends EventsTest
```
Comment (theofidry) [inline tests/Functional/Events_ServiceTest.php]: here as well

File: src/Events/ServiceContainerDispatcher.php (hunk 2 above)
Comment (wouterj) [inline src/Events/ServiceContainerDispatcher.php]: What do you mean? I can't typehint `array` any more specific, right?

File: tests/Functional/Events_ClassTest.php (hunk 3 above)
Comment (wouterj) [inline tests/Functional/Events_ClassTest.php]: Yeah, a bit weird personal convention :) Will remove the `_`

**Comment by wouterj** [comment]
> I left a few comments. Besides that WDYT of making those classes final? I think it's better to have everything as final by default.

I would agree with tagging everything `@final`, but I'm not in favor of marking them `final` in PHP. I don't think I should enforce people not to extend them, but I can tell them I don't help/support them if they do :)

**Comment by theofidry** [comment]
Fair for the final, in the end it's just a personal preference.

For the array type I meant being more specific with phpdoc blocks.

File: resources/docs/events.rst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+Then, pass the service ID in the ``Bundle::boot()`` method instead:
+
+.. code-block:: php
+
+    // ...
+    class AppBundle extends Bundle
+    {
+        public function boot()
+        {
+            User::observer('app.user_observer');
```
Comment (wouterj) [inline resources/docs/events.rst]: At first, I thought: Awesome idea!

However it turns out it's quite a bit harder than you think. So I created a new issue for it: https://github.com/wouterj/WouterJEloquentBundle/issues/52

**Comment by wouterj** [comment]
I don't think the PHPdoc will explain more than the method signature. I'll add all finals in a follow up PR after merging this one.

Thanks for your review!

File: resources/docs/events.rst (hunk 5 above)
Comment (theofidry) [inline resources/docs/events.rst]: Fair, I'll try to have a look at the end of the week, I don't recall when what is booted and when to be booted

**Comment by theofidry** [comment]
You're welcome. Do you have any ETA for a stable release or a beta/RC? Not urgent but I would prefer this lib to be somewhat stable when releasing the stable version of https://github.com/theofidry/AliceDataFixtures.

**Comment by wouterj** [comment]
@theofidry I'm hoping to release 0.4 stable this week. After that, I only have "nice to have" features left in the issue list afaics. So that means 0.5 might as well become 1.0. I'm not sure when I'll release 0.5/1.0, but it shouldn't take that long as there aren't much things left (other than Laravel 5.4 support).

The 0.x versions are already quite stable though.

**Pull request merged**